File: src/main/native/cpp/Simulation/SimDutyCycleEncoderManager/SimDutyCycleEncoderManager.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "OvertureLib/Simulation/SimDutyCycleEncoderManager/SimDutyCycleEncoderManager.h"
#include <iostream>

SimDutyCycleEncoderManager::SimDutyCycleEncoderManager() {

}

void SimDutyCycleEncoderManager::AddSimDutyCycleEncoderCandidate(
		OverDutyCycleEncoder *encoder) {
	dutyCycleEncodersToRegister.emplace_back(encoder);
}

void SimDutyCycleEncoderManager::RegisterSimDutyCycleEncoder(
		OverDutyCycleEncoder *encoder) {
	if (encoder == NULL || encoder == 0) {
		throw std::invalid_argument(
				"SimDutyCycleEncoderManager given null pointer!");
	}

	if (!this->CHNToDutyCycleEncoderNameMap.contains(
			encoder->GetSourceChannel())) {
		std::cout
				<< "SimDutyCycleEncoderManager Warning: Tried to register a DutyCycleEncoder for simulation("
				<< encoder->GetSourceChannel()
				<< ") that is not in the given map" << std::endl;
		return;
	}

	std::string encoderName =
			this->CHNToDutyCycleEncoderNameMap[encoder->GetSourceChannel()];

	if (this->registeredDutyCycleEncoders.contains(encoderName)) {
		std::cout
				<< "SimDutyCycleEncoderManager Warning: Tried to register a DutyCycleEncoder for simulation that was already registered"
				<< std::endl;
		return;
	}

	std::shared_ptr < nt::NetworkTable > ntable = ntInst.GetTable(encoderName);

	DutyCycleEncoderNTPair newDutyCycleEncoderPair;
	newDutyCycleEncoderPair.dutyCycleEncoder = std::make_shared
			< frc::sim::DutyCycleEncoderSim
			> (frc::sim::DutyCycleEncoderSim(*encoder));
	newDutyCycleEncoderPair.ntable = ntable;

	newDutyCycleEncoderPair.dutyCycleEncoder->SetConnected(true);
	this->registeredDutyCycleEncoders[encoderName] = std::move(
			newDutyCycleEncoderPair);
}

void SimDutyCycleEncoderManager::Init(
		const std::map<unsigned int, NTDutyCycleEncoderName> CHNToDutyCycleEncoderNameMap) {
	this->CHNToDutyCycleEncoderNameMap = CHNToDutyCycleEncoderNameMap;

	std::for_each(dutyCycleEncodersToRegister.begin(),
			dutyCycleEncodersToRegister.end(),
			std::bind(&SimDutyCycleEncoderManager::RegisterSimDutyCycleEncoder,
					this, std::placeholders::_1));
	initialized = true;
}

void SimDutyCycleEncoderManager::Update() {
	if (!initialized) {
		return;
	}

	for (auto encoderIterator = this->registeredDutyCycleEncoders.begin();
			encoderIterator != this->registeredDutyCycleEncoders.end();
			encoderIterator++) {
		DutyCycleEncoderNTPair pair = encoderIterator->second;

		std::shared_ptr < nt::NetworkTable > ntable = pair.ntable;
		std::shared_ptr < frc::sim::DutyCycleEncoderSim > encoder =
				pair.dutyCycleEncoder;

		// units::turn_t position = units::turn_t();
		encoder->SetAbsolutePosition(
				ntable->GetEntry("cancoder_position").GetDouble(0));
		encoder->SetConnected(ntable->GetEntry("cancoder_position").Exists());
	}

	ntInst.Flush();
}

SimDutyCycleEncoderManager& SimDutyCycleEncoderManager::GetInstance() {
	static SimDutyCycleEncoderManager instance;
	return instance;
}

File: src/main/java/com/overture/lib/simulation/SimDutyCycleEncoderManager.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package com.overture.lib.simulation;

import com.overture.lib.sensors.OverDutyCycleEncoder;
import edu.wpi.first.networktables.NetworkTable;
import edu.wpi.first.networktables.NetworkTableInstance;
import edu.wpi.first.wpilibj.simulation.DutyCycleEncoderSim;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Bridges every {@link OverDutyCycleEncoder} to an external physics simulation over NetworkTables.
 */
public final class SimDutyCycleEncoderManager {
  private static final SimDutyCycleEncoderManager instance = new SimDutyCycleEncoderManager();

  private final NetworkTableInstance ntInst = NetworkTableInstance.getDefault();

  private Map<Integer, String> chnToDutyCycleEncoderNameMap = new HashMap<>();
  private final Map<String, DutyCycleEncoderNTPair> registeredDutyCycleEncoders = new HashMap<>();
  private final List<OverDutyCycleEncoder> dutyCycleEncodersToRegister = new ArrayList<>();
  private boolean initialized = false;

  private static final class DutyCycleEncoderNTPair {
    private NetworkTable ntable;
    private DutyCycleEncoderSim dutyCycleEncoder;
  }

  private SimDutyCycleEncoderManager() {}

  /**
   * Returns the singleton instance.
   *
   * @return the sim duty cycle encoder manager
   */
  public static SimDutyCycleEncoderManager getInstance() {
    return instance;
  }

  /**
   * Queues an encoder to be registered once {@link #init(Map)} runs.
   *
   * @param encoder the encoder to register
   */
  public void addSimDutyCycleEncoderCandidate(OverDutyCycleEncoder encoder) {
    dutyCycleEncodersToRegister.add(encoder);
  }

  /**
   * Registers every queued encoder whose channel appears in the given map.
   *
   * @param chnToDutyCycleEncoderNameMap maps DIO channel to the NetworkTables name of the simulated
   *     encoder
   */
  public void init(Map<Integer, String> chnToDutyCycleEncoderNameMap) {
    this.chnToDutyCycleEncoderNameMap = chnToDutyCycleEncoderNameMap;

    dutyCycleEncodersToRegister.forEach(this::registerSimDutyCycleEncoder);
    initialized = true;
  }

  private void registerSimDutyCycleEncoder(OverDutyCycleEncoder encoder) {
    if (encoder == null) {
      throw new IllegalArgumentException("SimDutyCycleEncoderManager given null pointer!");
    }

    if (!chnToDutyCycleEncoderNameMap.containsKey(encoder.getSourceChannel())) {
      System.out.println(
          "SimDutyCycleEncoderManager Warning: Tried to register a DutyCycleEncoder for"
              + " simulation("
              + encoder.getSourceChannel()
              + ") that is not in the given map");
      return;
    }

    String encoderName = chnToDutyCycleEncoderNameMap.get(encoder.getSourceChannel());

    if (registeredDutyCycleEncoders.containsKey(encoderName)) {
      System.out.println(
          "SimDutyCycleEncoderManager Warning: Tried to register a DutyCycleEncoder for simulation"
              + " that was already registered");
      return;
    }

    DutyCycleEncoderNTPair newDutyCycleEncoderPair = new DutyCycleEncoderNTPair();
    newDutyCycleEncoderPair.dutyCycleEncoder = new DutyCycleEncoderSim(encoder);
    newDutyCycleEncoderPair.ntable = ntInst.getTable(encoderName);

    newDutyCycleEncoderPair.dutyCycleEncoder.setConnected(true);
    registeredDutyCycleEncoders.put(encoderName, newDutyCycleEncoderPair);
  }

  /** Pushes the simulated sensor state into every encoder. Call this periodically. */
  public void update() {
    if (!initialized) {
      return;
    }

    for (DutyCycleEncoderNTPair pair : registeredDutyCycleEncoders.values()) {
      NetworkTable ntable = pair.ntable;
      DutyCycleEncoderSim encoder = pair.dutyCycleEncoder;

      encoder.set(ntable.getEntry("cancoder_position").getDouble(0));
      encoder.setConnected(ntable.getEntry("cancoder_position").exists());
    }

    ntInst.flush();
  }
}
